### ISSUE TITLE
get user and email from $HOME/.gitconfig

### ISSUE THREAD

**Issue by ghjagus** [isomorphic-git/isomorphic-git#236]
this is the code, and the console.log() prints undefined。
But i can get the email value use 'git config user.email' in shell
  
```js
async function f () {
    let value = await git.config(Object.assign({
        path: 'user.email',
    }, repo));

    console.log(value); 
}
```

**Comment by wmhilton** [comment]
Is the value defined locally, in `.git/config` or globally in `~/.gitconfig`? Right now `isomorphic-git` can only read the local repo config file, not the one in the home directory. (I haven't figured out what the "home directory" would be in a browser environment.)

**Comment by TomasHubelbauer** [comment]
Maybe this would be another candidate for a plugin? "Environment provider" kinda thing, would have a mechanism for resolving the home path / on the fly providing home entries.

**Comment by wmhilton** [comment]
> Maybe this would be another candidate for a plugin? "Environment provider" kinda thing, would have a mechanism for resolving the home path / on the fly providing home entries.

Maybe? But I feel like it would be beneficial for the whole BrowserFS ecosystem if some convention, like a top-level folder called '~' was chosen as a standard config location. How are you handling it @jcubic ?

**Comment by jcubic** [comment]
I'm not handling config files yet, I only have `git login` command that set the object and use it in commit and push.

> I haven't figured out what the "home directory" would be in a browser environment.

What about root directory, `/.gitconfig`. When I'm cloning repo it's always in `/<repo>` and root directory is usually empty, unless someone is write a file there using editor.

**Comment by wmhilton** [comment]
> What about root directory, `/.gitconfig`

I like that idea, of putting them at the root directory level. Would work nicely for `.babelrc` and `.editorconfig` etc. Makes it obvious to the user where they are.

**Comment by wmhilton** [comment]
Basically, we shim `os.homedir()` to return '/' on BrowserFS.

**Comment by ghjagus** [comment]
> Is the value defined locally, in .git/config or globally in ~/.gitconfig? Right now isomorphic-git can only read the local repo config file, not the one in the home directory. (I haven't figured out what the "home directory" would be in a browser environment.)

My env is nodejs in windows 7, And i only set the value globally.

**Comment by jcubic** [comment]
what about:

```javascript
git.plugins.set('.gitconfig', '/some/path/.gitconfig');
```
or

```javascript
// browser
git.plugins.set('home', '/');
// node
git.plugins.set('home', require('os').homedir());
```

that's way you don't need `~` directory but if user want he can use it (eg. on node)

**Comment by jcubic** [comment]
In my git terminal I don't get any use from credential plugin that I need to get username and email from user, right now I have `login` command that get all those data from user so If I use credentials plugin I would need to ask twice first for user/email and second for username and password. (I've didn't merged credential branch yet - with cred plugin - I've didn't work on my terminal for a while).

What about if credential plugin ask for email/fullname and if there is global `.gitconfig` it use that file.

**Comment by brettz9** [comment]
FWIW, I'm using the following workaround for Node:

```js
const parseGitConfig = require('parse-git-config');
const getGitConfigPath = require('git-config-path');

exports.getGlobalGitAuthorInfo = async () => {
  const globalGitConfigPath = getGitConfigPath('global');
  const parsedConfig = await parseGitConfig({
    path: globalGitConfigPath
  });
  return parsedConfig;
};
```

I then use it by trying a regular commit without any author info supplied to `commit`, and if that errs, I use the following (with try-catches omitted):

```js
const {user: {name: globalGitAuthorName, email: globalGitAuthorEmail}} = await getGlobalGitAuthorInfo();
 
await commit({repositoryPath, author: {
    name: globalGitAuthorName,
    email: globalGitAuthorEmail
}});
```

**Comment by kdankert** [comment]
Are there any updates on this, I would really like to use isomorphic git as an alternative to nodegit, which extremely slows down my CI/CD jobs, because of the native build.....
Isomorphic-git meets all my other requirements, except reading from the global config file. 
I don't really want to use more libraries than I have to, so i would like to skip that workaround.

**Comment by SamVerschueren** [comment]
What if you could just provide a `filepath` property instead to read the config from?

This will just read from the local config as it does right now

```ts
await git.getConfig({
   fs,
   gitdir: '.../.git',
   path: 'user.email'
});
```

This will read from the `~/.gitconfig` file instead.

```ts
await git.getConfig({
   fs,
   gitdir: '.../.git',
   filepath: path.join(os.homedir(), '.gitconfig'),
   path: 'user.email'
});
```

Feels like the most flexible way of offering the ability to read/write to config files in different locations. For instance, `git config` also has a `--file` option where someone can provide a custom location, this will also work for that use case.

I can send a PR if this is the path we want to take.

**Comment by SamVerschueren** [comment]
After implementing this internally (by copy-pasting some files out of `isomorphic-git`), I don't think my proposed solution is what we need/want. So here's a new proposal.

Add the following properties to the `getConfig` options
- `local: boolean`: If set explicitly, read the value from the local config file only (`gitdir` is mandatory in this case)
- `global: boolean`: If set explicitly, read the value from the global config file (`gitdir` is optional in this case, `os` is mandatory)
- `filepath: string`: If set, read the value from the provided filepath (`gitdir` is optional in this case)
- `os: object`: An object which has a `homedir()` method just like Node.js `os.homedir()`. Browser can provide a custom implementation, Node.js can just provide `require('os')`.

Here are some scenario's

```ts
await git.getConfig({
   fs,
   os,
   gitdir: '.../.git',
   path: 'user.name'
});
//=> Not found in local config. Read from global config because `os` is provided.
```

```ts
await git.getConfig({
   fs,
   os,
   path: 'user.name'
});
//=> `gitdir` not provided, read from global config directly because it means we are not in a git directory
```

```ts
await git.getConfig({
   fs,
   os,
   gitdir: '.../.git',
   path: 'user.name',
   local: true
});
//=> Not found in local config. Does not read from global config because `local` explicitly set
```

```ts
await git.getConfig({
   fs,
   os,
   gitdir: '.../.git',
   path: 'user.name',
   global: true
});
//=> Skipping local config, reads directly from global config.
```

```ts
await git.getConfig({
   fs,
   os,
   path: 'user.name',
   filepath: 'path/to/custom-config'
});
//=> Read from `custom-config` file
```

I think the same kinda goes for `getConfigAll` and `setConfig`.

### Argument checks
1. `local`, `global` or `filepath` are mutually exclusive, so it should throw if more than 1 of these are provided.
2. If `local` is provided, `gitdir` should be provided as well.
3. If `local` or `filepath` are not provided, `os` should be provided.

**Comment by roddc** [comment]
Is this feature still in progress?

**Comment by jcubic** [comment]
@roddc No, unless someone will want to implement this. This project is driven exclusively by contributors. It doesn't have the main author that does all the hard work.
